compiler/fir/providers/src/org/jetbrains/kotlin/fir/scopes/impl/FirTypeIntersectionScopeContext.kt

```diff
@@ -372,81 +347,89 @@ class FirTypeIntersectionScopeContext(
-                    mostSpecific.fir.name
-                ),
-                overrides
-            )
-        val mostSpecificFunction = mostSpecific.fir
+        val key = mostSpecific.first() as FirNamedFunctionSymbol
+        val keyFir = key.fir
+        val callableId = CallableId(
+            dispatchReceiverType.classId ?: keyFir.dispatchReceiverClassOrNull()?.classId!!,
+            keyFir.name
+        )
+        val newSymbol = FirIntersectionOverrideFunctionSymbol(callableId, overrides)
         FirFakeOverrideGenerator.createCopyForFirFunction(
-            newSymbol,
-            mostSpecificFunction, session, FirDeclarationOrigin.IntersectionOverride,
-            mostSpecificFunction.isExpect,
-            newDispatchReceiverType = dispatchReceiverType,
+            newSymbol, keyFir, session, FirDeclarationOrigin.IntersectionOverride, keyFir.isExpect,
             newModality = newModality,
             newVisibility = newVisibility,
+            newDispatchReceiverType = dispatchReceiverType,
+            newReturnType = if (!forClassUseSiteScope) intersectReturnTypes(mostSpecific) else null,
         ).apply {
-            originalForIntersectionOverrideAttr = mostSpecific.fir
+            originalForIntersectionOverrideAttr = keyFir
         }
         return newSymbol
     }
 
-    private fun createIntersectionOverride(
-        mostSpecific: FirPropertySymbol,
+    private fun createIntersectionOverrideProperty(
+        mostSpecific: Collection<FirCallableSymbol<*>>,
         overrides: Collection<FirCallableSymbol<*>>,
         newModality: Modality?,
         newVisibility: Visibility,
     ): FirPropertySymbol {
+        val key = mostSpecific.first() as FirPropertySymbol
+        val keyFir = key.fir
         val callableId = CallableId(
-            dispatchReceiverType.classId ?: mostSpecific.dispatchReceiverClassOrNull()?.classId!!,
-            mostSpecific.fir.name
+            dispatchReceiverType.classId ?: keyFir.dispatchReceiverClassOrNull()?.classId!!,
+            keyFir.name
         )
         val newSymbol = FirIntersectionOverridePropertySymbol(callableId, overrides)
-        val mostSpecificProperty = mostSpecific.fir
         FirFakeOverrideGenerator.createCopyForFirProperty(
-            newSymbol, mostSpecificProperty, session, FirDeclarationOrigin.IntersectionOverride,
+            newSymbol, keyFir, session, FirDeclarationOrigin.IntersectionOverride,
             newModality = newModality,
             newVisibility = newVisibility,
             newDispatchReceiverType = dispatchReceiverType,
+            // If any of the properties are vars and the types are not equal, these declarations are conflicting
+            // anyway and their uses should result in an overload resolution error.
+            newReturnType = if (!forClassUseSiteScope && !mostSpecific.any { (it as FirPropertySymbol).fir.isVar })
+                intersectReturnTypes(mostSpecific)
+            else
+                null,
```

Comment:
Speaking of this case - FIR currently does not report that error (diagnostics/tests/smartCasts/intersectionScope/propertiesConflict.kt):

```
interface A {
    val foo: Any?
}

interface C: A {
    override val foo: String
}
interface B: A {
    override var foo: String?
}

fun test(a: A) {
    if (a is B && a is C) {
        a.foo = "" // K1: OVERLOAD_RESOLUTION_AMBIGUITY; K2: OK
        a.foo = null // K1: OVERLOAD_RESOLUTION_AMBIGUITY; K2: OK
        a.foo // K1: OVERLOAD_RESOLUTION_AMBIGUITY; K2: OK
    }
}
```

What I'm wondering here is whether instead there should be a CAST_NEVER_SUCCEEDS on `a is C`, since it shouldn't be possible to have a type that implements both interfaces due to conflicting declarations?

Reply:
Not sure it would be computationally efficient (we would have to check all the declarations there).
What might be useful is to warn/forbid using such ill-formed intersection scope's declarations.
But that's a different story, I think.